### ISSUE TITLE
Serialising patch requests not working correctly

### ISSUE THREAD

**Issue by andye2004** [BlackPepperSoftware/bowman#21]
Not really sure whether you would consider this a bug or a feature but I thought I'd raise it in case I can save other people some pain. I am using 0.5.0-RELEASE2 which I upgraded to so that we could use the new PATCH functionality.

The issue I am seeing is this:

1. I have an entity being served up by a Spring Data REST service that contains a linked resource.
2. I modify the entity in a web-app to update the linked resource to point to a different resource.
3. When the request is serialised to send to the SDR service it always serialises the 'old' linked resource.

After some investigation this appears to be because of a combination of two things:

1. When the resource setter method is called it does actually update the underlying instance BUT the linked resources contained in the LinkedResourceMethodHandler are never updated.
2. The handler chain has the LinkedResourceMethodHandler in the chain BEFORE the getter handler which means that it returns the resource from the list in the LinkedResourceMethodHandler which is the original linked resource.

Would be interested to get your thoughts on this and if you had any ideas for a possible work around / fix.

I am aware of #2 but that obviously relates to PUT operations rather than PATCH.

Thanks, Andy.

**Comment by andye2004** [comment]
Been looking at this a little bit more and I do consider it a bug mainly because there is the LinkedResourceUriSerializer class in the JacksonClientModule class file. This class is trying to write out the URI for the linked resource on any serialisation of the entity but as the instance has been updated I think it is actually writing out the wrong one.

Also, after looking more closely at things I was wondering about the possibility of using LinkedResource annotation on setters as well as getters? If we did this then the LinkedResourceMethodHandler would be triggered for setters as well, then in the invoke method we could look at whether it was a setter or getter before deciding what to do? If it were a setter method being called we could replace the value in the map instead of calling the service for the result?

Happy to do the work on this if someone can provide a little guidance?

**Comment by hdpe** [comment]
Hi, thanks very much for reporting this!

I agree with you - it's definitely a bug, and a pretty massive one at that! Your analysis of the problem seems spot on. I don't think it's PATCH-specific - basically any time you mutate a `@LinkedResource` property of a proxied response and re-serialise, the original value is marshalled instead, because the handler uses the version that's cached within it.

I'm not sure the best way to fix this yet but I think you're on to something with your second comment - I'm not sure it makes sense to split up handlers based on whether they're getting or setting - you kind of want them to share the same instance state regardless. It may well make more sense just to define handlers based on what annotation (or lack of) that they're targeting. Looks like quite a big refactor though: it's certainly making me question whether my splitting of the old massive monolithic `MethodHandler` into a chain of little ones was necessarily a good idea!

Would you be interested in raising a PR to address all this? Once deciding on a strategy if need be?

Regarding a temporary workaround - all I can think of is wondering whether there's some way you can 'clone' the response state before re-submitting. Maybe you could use some JavaBeans-property-aware lib like `BeanUtils`, or maybe you could use Jackson but with a new `ObjectMapper` that doesn't register the Bowman module?

**Comment by andye2004** [comment]
Hi Ryan, I actually have a working patch in test at the moment that seems to do the trick. Without copying the whole thing here, the salient parts are:

```java
class LinkedResourceMethodHandler extends AbstractContentDelegatingMethodHandler {
.
.
.
    @Override
    public boolean supports(Method method) {
        return method.isAnnotationPresent(LinkedResource.class);
    }
	
    @Override
    public Object invoke(Object self, Method method, Method proceed, Object[] args)
            throws InvocationTargetException, IllegalAccessException {

        if (isSetter(method)) {
            super.invoke(self, method, proceed, args);
            linkedResourceResults.put(getGetterFromSetter(method).getName(), args[0]);
        }
        else {
            if (!linkedResourceResults.containsKey(method.getName())) {
                linkedResourceResults.put(method.getName(), resolveLinkedResource(self, method, proceed, args));
            }
        }
        return linkedResourceResults.get(method.getName());
    }

    private boolean isSetter(Method method) {
            return Arrays.stream(getContentBeanInfo().getPropertyDescriptors())
                    .map(PropertyDescriptor::getWriteMethod)
                    .anyMatch(method::equals);
    }

    private Method getGetterFromSetter(Method method) {
            return Arrays.stream(getContentBeanInfo().getPropertyDescriptors())
                    .filter(pd -> method.equals(pd.getWriteMethod()))
                    .collect(Collectors.toList()).get(0).getReadMethod();
    }
.
.
.
}
```
There are a couple of other really small changes that are required to support this but ommitted here for brevity. I'll create a fork later and apply the patch to and raise a PR and we can discuss it and modify as required.

**Comment by hdpe** [comment]
Brilliant, thanks!

**Comment by hdpe** [comment]
Fixed by #22
